src/lib/parquet/src/Flow/Parquet/ParquetFile/Data/RLEBitPackedHybrid.php

```diff
@@ -112,7 +111,115 @@ public function decodeRLE(BinaryReader $reader, int $bitWidth, int $intVar, int
         }
     }
 
-    private function debugLog(int $iteration, int $varInt, bool $isRle, int $bitWidth, DataSize $length, BinaryReader $reader, array $output) : void
+    /**
+     * @param array<int> $values
+     */
+    public function encodeBitPacked(BinaryWriter $writer, int $bitWidth, array $values) : void
+    {
+        $numGroups = (int) \ceil(\count($values) / 8.0);
+        $varInt = ($numGroups << 1) | 1;
+
+        $writer->writeVarInts32([$varInt]);
+
+        $buffer = 0;
+        $bitsInBuffer = 0;
+        $bytes = [];
+
+        foreach ($values as $value) {
+            $buffer |= ($value << $bitsInBuffer);
+            $bitsInBuffer += $bitWidth;
+
+            while ($bitsInBuffer >= 8) {
+                $bytes[] = $buffer & 0xFF;
+                $buffer >>= 8;
+                $bitsInBuffer -= 8;
+            }
+        }
+
+        // Write any remaining bits in the buffer
+        if ($bitsInBuffer > 0) {
+            $bytes[] = $buffer & 0xFF;
+        }
+
+        $writer->writeBytes($bytes);
+    }
+
+    /**
+     * @param array<int> $values
+     */
+    public function encodeHybrid(BinaryWriter $writer, array $values) : void
+    {
+        $bitWidth = BitWidth::fromArray($values);
+
+        $rleBuffer = [];
+        $bitPackedBuffer = [];
+
+        $previousValue = null;
+
+        foreach ($values as $value) {
+            if ($previousValue === null) {
+                $previousValue = $value;
+                $rleBuffer[] = $value;
+
+                continue;
+            }
+
+            // we always bit-pack a multiple of 8 values at a time, so we only store the number of values / 8
+            if (\count($bitPackedBuffer) > 0 && \count($bitPackedBuffer) < 8) {
+                $bitPackedBuffer[] = $value;
+
+                continue;
+            }
+
+            if ($previousValue === $value) {
+                $rleBuffer[] = $value;
+            } else {
+                if (\count($rleBuffer) >= 8) {
+                    $this->encodeRLE($writer, $bitWidth, $rleBuffer);
+                    $rleBuffer = [];
+                }
+
+                $bitPackedBuffer = \array_merge($bitPackedBuffer, $rleBuffer);
+                $bitPackedBuffer[] = $value;
+                $rleBuffer = [];
+            }
+
+            $previousValue = $value;
+        }
+
+        if (\count($rleBuffer) > 8) {
+            $this->encodeRLE($writer, $bitWidth, $rleBuffer);
+            $rleBuffer = [];
+        }
+
+        if (\count($bitPackedBuffer)) {
+            if (\count($rleBuffer)) {
```

Comment:
Probably can be removed

Reply:
it can't, if there is a situation where rleBuffer is not full, there are no more values and bitPackedBuffer is also not empty, in that case, we would lose those last values.